rmf_rviz_plugin/src/RmfPanel.cpp

```diff
@@ -13,404 +13,758 @@
+bool RmfPanel::waypoint_has_workcell(std::string waypoint_name,
+                                     GraphInfo &graph_info) 
 {
-  std::lock_guard<std::mutex> lock(_mutex);
-  _delivery_dropoff = value;
-  Q_EMIT configChanged();
+  auto idx = graph_info.keys.find(waypoint_name);
+  if (idx == graph_info.keys.end()) 
+  {
+    RCLCPP_ERROR(_node->get_logger(),
+                 "Provided graph does not have this waypoint.");
+    return false;
+  }
+  if (graph_info.workcell_names.find(idx->second) ==
+      graph_info.workcell_names.end()) \
+  {
+    // Workcell does not exist at this waypoint
+    return false;
+  } else {
+    return true;
+  }
 }
 
-void RmfPanel::set_delivery_robot(const QString& value)
+RmfPanel::RmfPanel(QWidget *parent) : rviz_common::Panel(parent) 
 {
-  std::lock_guard<std::mutex> lock(_mutex);
-  _delivery_robot = value;
-  Q_EMIT configChanged();
+  _node = std::make_shared<rclcpp::Node>("rmf_panel");
+  _update_timer = new QTimer(this);
+  _update_timer->start(1000); // Update clock running at 1Hz
+
+  create_layout();
+  initialize_subscribers(_node);
+  initialize_publishers(_node);
+  initialize_qt_connections();
+  initialize_models();
+
+  _thread = std::thread([&]() { rclcpp::spin(_node); });
+
+  _has_loaded = true;
 }
 
-// Loop set functions
-void RmfPanel::set_loop_task_id(const QString& value)
+RmfPanel::~RmfPanel() 
 {
-  std::lock_guard<std::mutex> lock(_mutex);
-  _loop_task_id = value;
-  Q_EMIT configChanged();
+  if (_has_loaded) 
+  {
+    _thread.join();
+    rclcpp::shutdown();
+  }
 }
 
-void RmfPanel::set_loop_start(const QString& value)
+// Load config
+void RmfPanel::load(const rviz_common::Config &config) 
 {
-  std::lock_guard<std::mutex> lock(_mutex);
-  _loop_start = value;
-  Q_EMIT configChanged();
+  rviz_common::Panel::load(config);
 }
 
-void RmfPanel::set_loop_finish(const QString& value)
+// Save config
+void RmfPanel::save(rviz_common::Config config) const 
 {
-  std::lock_guard<std::mutex> lock(_mutex);
-  _loop_finish = value;
-  Q_EMIT configChanged();
+  rviz_common::Panel::save(config);
 }
 
-void RmfPanel::set_loop_robot(const QString& value)
+// Q_SLOTS
+// Actions
+void RmfPanel::queue_delivery() 
 {
-  std::lock_guard<std::mutex> lock(_mutex);
-  _loop_robot = value;
+  std::string start = _start_waypoint_selector->currentText().toStdString();
+  std::string end = _end_waypoint_selector->currentText().toStdString();
+  Delivery delivery;
+
+  // If either start or end are empty, the delivery should probably not be queued
+  if (start.empty() || end.empty())
+  {
+    RCLCPP_INFO(_node->get_logger(), "Waypoint input is empty string; Delivery not queued in plan.");
+    return;
+  }
+
+  delivery.task_id = generate_task_uuid(3);
+  delivery.pickup_place_name = start;
+  delivery.dropoff_place_name = end;
+
+  auto delivery_time = _time_selector->time();
+
+  int insertPos = 0;
+  for (auto it = _queued_deliveries.begin(); it != _queued_deliveries.end();
```

Comment:
since we have `pop_delivery()` locking and operating on a timer, we should lock before iterating over `_queued_deliveries`

Reply:
:+1: 

it seems [`std::lock_guard`](https://en.cppreference.com/w/cpp/thread/lock_guard) releases the mutex after leaving the local scope, so there is no need to manually release a lock.